packages/iov-bns/src/bnsconnection.ts

```diff
@@ -192,23 +196,33 @@ export class BnsConnection implements BcpAtomicSwapConnection {
     // can be undefined as we cannot guarantee it assigned before the caller unsubscribes from the stream
     let blockHeadersSubscription: Subscription | undefined;
 
-    const firstEvent: BcpBlockInfo = { state: BcpTransactionState.Pending };
-    let lastEventSent: BcpBlockInfo = firstEvent;
-    const blockInfoProducer = new DefaultValueProducer<BcpBlockInfo>(firstEvent, {
+    const firstEvent: BlockInfo = { state: TransactionState.Pending };
+    let lastEventSent: BlockInfo = firstEvent;
+    const blockInfoProducer = new DefaultValueProducer<BlockInfo>(firstEvent, {
       onStarted: async () => {
         try {
-          // we utilize liveTx to implement a _search or watch_ mechanism since we do not know
-          // if the transaction is already committed when the producer is started
-          const searchResult = await toListPromise(this.liveTx({ id: transactionId }), 1);
```

Comment:
Hmmm... I see the point of using streams internally, but do we have to expose it?

For example, if we use something like this (from the test code):
```
      const pendingResult = toListPromise(connection.waitForTransaction(transactionId), 1);

      // send transaction
      await connection.postTx(liskCodec.bytesToPost(signed));

      // wait for transaction
      const result = (await pendingResult)[0];
```

we will block forever and never unsubscribe

Reply:
To make it easy for the client to use the library properly, I would propose using the streams and unsubscribe *internally* as you did here, but exposing with a timeout parameter, where it would unsubscribe all (potentially polling) subscriptions and reject the promise.

eg. `const result = await connection.waitForTransaction(transactionId, 15)` will cancel all subscriptions and reject the promise after 15 seconds. 

If you want to expose all this power to the users, assuming they want real reactive, maybe leave this as is, and add `waitForTransactionWithTimeout`, which returns a promise, and would be the simple API for the non power-user